src/app/clusters/smoke-co-alarm-server/smoke-co-alarm-server.cpp

```diff
@@ -72,6 +61,101 @@ bool SmokeCoAlarmServer::SetExpressedState(EndpointId endpointId, ExpressedState
     return success;
 }
 
+bool SmokeCoAlarmServer::AutoSetExpressedState(EndpointId endpointId,
+                                               std::array<ExpressedStateEnum, kPriorityOrderLength> priorityOrder)
+{
+    ExpressedStateEnum currentExpressedState = ExpressedStateEnum::kNormal;
+
+    for (ExpressedStateEnum priority : priorityOrder)
+    {
+        switch (priority)
+        {
+        case ExpressedStateEnum::kSmokeAlarm: {
+            AlarmStateEnum currentSmokeAlarm;
+            VerifyOrReturnValue(GetSmokeState(endpointId, currentSmokeAlarm), false);
+            if (currentSmokeAlarm != AlarmStateEnum::kNormal)
```

Comment:
are these conditionals needed? We seem to exit the loop when currentSmokeAlarm is not normal anyway.